File: frontend/src/services/distributionService.ts
// services/distributionService.ts

import { TimeGranularity } from "../types/TimeGranularity"
import { MedicinalUnitMode } from "../types/MedicinalUnitMode"
import {MedicineProductInfo} from "../types/MedicineProductInfo";

// ======= DTOs: Sankey Diagram =======

export type DistributionSankeyRequest = {
    medicinalProductIds: number[]
    registrationNumbers: string[]
    dateFrom: string // "yyyy-MM"
    dateTo: string // "yyyy-MM"
    medicinalUnitMode: MedicinalUnitMode
}

export type SankeyNodeDto = {
    id: string
    label: string
}

export type SankeyLinkDto = {
    source: string
    target: string
    value: number
}

export type DistributionSankeyResponse = {
    includedMedicineProducts: MedicineProductInfo[]
    ignoredMedicineProducts: MedicineProductInfo[]
    dateFrom: string
    dateTo: string
    medicinalUnitMode: MedicinalUnitMode
    nodes: SankeyNodeDto[]
    links: SankeyLinkDto[]
}

export async function fetchDistributionSankeyDiagram(
    request: DistributionSankeyRequest
): Promise<DistributionSankeyResponse> {
    const response = await fetch("/api/distribution/sankey-diagram", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request)
    })

    if (!response.ok) {
        throw new Error("Failed to fetch data for distribution Sankey diagram")
    }

    return response.json()
}

// ======= DTOs: Time Series =======

export type DistributionTimeSeriesRequest = {
    medicinalProductIds: number[]
    registrationNumbers: string[]
    dateFrom: string // "yyyy-MM"
    dateTo: string // "yyyy-MM"
    medicinalUnitMode: MedicinalUnitMode
    timeGranularity: TimeGranularity
}

export type DistributionFlowEntry = {
    source: string
    target: string
    value: number
}

export type DistributionTimeSeriesPeriodEntry = {
    period: string
    flows: DistributionFlowEntry[]
}

export type DistributionTimeSeriesResponse = {
    includedMedicineProducts: MedicineProductInfo[]
    ignoredMedicineProducts: MedicineProductInfo[]
    dateFrom: string
    dateTo: string
    medicinalUnitMode: MedicinalUnitMode
    timeGranularity: TimeGranularity
    series: DistributionTimeSeriesPeriodEntry[]
}

export async function fetchDistributionTimeSeries(
    request: DistributionTimeSeriesRequest
): Promise<DistributionTimeSeriesResponse> {
    const response = await fetch("/api/distribution/time-series", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request)
    })

    if (!response.ok) {
        throw new Error("Failed to fetch data for distribution time series")
    }

    return response.json()
}

File: src/main/kotlin/cz/machovec/lekovyportal/api/service/DistributionService.kt
package cz.machovec.lekovyportal.api.service

import DistributionSankeyRequest
import DistributionSankeyResponse
import DistributionTimeSeriesRequest
import DistributionTimeSeriesResponse
import cz.machovec.lekovyportal.api.calculations.DoseUnitConverterFactory
import cz.machovec.lekovyportal.api.calculations.distribution.SankeyDiagramAssembler
import cz.machovec.lekovyportal.api.calculations.distribution.TimeSeriesAssembler
import cz.machovec.lekovyportal.api.model.enums.MedicinalUnitMode
import cz.machovec.lekovyportal.api.model.mpd.MedicinalProductIdentificators
import cz.machovec.lekovyportal.core.domain.mpd.MpdMedicinalProduct
import cz.machovec.lekovyportal.core.repository.distribution.*
import cz.machovec.lekovyportal.core.repository.mpd.MpdMedicinalProductRepository
import org.springframework.stereotype.Service
import java.math.BigDecimal

@Service
class DistributionService(
    private val mahRepo: DistFromMahsRepository,
    private val distRepo: DistFromDistributorsRepository,
    private val pharmRepo: DistFromPharmaciesRepository,
    private val medicinalProductRepo: MpdMedicinalProductRepository,
    private val converterFactory: DoseUnitConverterFactory,
    private val sankeyAssembler: SankeyDiagramAssembler,
    private val timeSeriesAssembler: TimeSeriesAssembler
) {

    /** PUBLIC API **/

    fun getSankeyDiagram(req: DistributionSankeyRequest): DistributionSankeyResponse {

        // Step 1: Load and filter medicinal products by DDD compatibility
        val (included, ignored) = loadAndFilterProducts(
            req.medicinalProductIds,
            req.registrationNumbers,
            req.medicinalUnitMode
        )

        val dddPerProduct = included.associate { it.id!! to (it.dailyDosePackaging ?: BigDecimal.ZERO) }
        val productIds    = included.mapNotNull { it.id }

        // Step 2: Fetch aggregated rows from all three reporting sources
        val (fromY, fromM) = req.dateFrom.split("-").let { it[0].toInt() to it[1].toInt() }
        val (toY,   toM)   = req.dateTo.split("-").let { it[0].toInt() to it[1].toInt() }

        val mahRows   = mahRepo.getAggregateProductMovementCounts(productIds, fromY, fromM, toY, toM)
        val distRows  = distRepo.getAggregateDistributorProductMovementCounts(productIds, fromY, fromM, toY, toM)
        val pharmRows = pharmRepo.getAggregatePharmacyProductDispenseCounts(productIds, fromY, fromM, toY, toM)

        // Step 3: Convert units (PACKAGES ↔ DDD) and assemble nodes + links
        val converter      = converterFactory.of(req.medicinalUnitMode)
        val (nodes, links) = sankeyAssembler.assemble(mahRows, distRows, pharmRows, converter, dddPerProduct)

        // Step 4: Build response DTO
        return DistributionSankeyResponse(
            includedMedicineProducts = included.toDto(),
            ignoredMedicineProducts  = ignored.toDto(),
            dateFrom          = req.dateFrom,
            dateTo            = req.dateTo,
            medicinalUnitMode = req.medicinalUnitMode,
            nodes             = nodes,
            links             = links
        )
    }

    fun getTimeSeries(req: DistributionTimeSeriesRequest): DistributionTimeSeriesResponse {

        // Step 1: Load and filter medicinal products by DDD compatibility
        val (included, ignored) = loadAndFilterProducts(
            req.medicinalProductIds,
            req.registrationNumbers,
            req.medicinalUnitMode
        )

        val dddPerProduct = included.associate { it.id!! to (it.dailyDosePackaging ?: BigDecimal.ZERO) }
        val productIds    = included.mapNotNull { it.id }

        // Step 2: Fetch raw monthly rows for all reporters
        val mahRows   = mahRepo.getMonthlyMahProductMovementCounts(productIds)
        val distRows  = distRepo.getMonthlyDistributorProductMovementCounts(productIds)
        val pharmRows = pharmRepo.getMonthlyPharmacyProductDispenseCounts(productIds)

        // Step 3: Convert units + assemble per-period flows
        val converter = converterFactory.of(req.medicinalUnitMode)
        val series    = timeSeriesAssembler.assemble(
            mahRows, distRows, pharmRows,
            req.timeGranularity, converter, dddPerProduct
        )

        // Step 4: Build response DTO
        return DistributionTimeSeriesResponse(
            includedMedicineProducts = included.toDto(),
            ignoredMedicineProducts  = ignored.toDto(),
            dateFrom          = req.dateFrom,
            dateTo            = req.dateTo,
            medicinalUnitMode = req.medicinalUnitMode,
            timeGranularity   = req.timeGranularity,
            series            = series
        )
    }

    /** PRIVATE UTILITIES **/

    private fun loadAndFilterProducts(
        ids: List<Long>,
        regNumbers: List<String>,
        unitMode: MedicinalUnitMode
    ): Pair<List<MpdMedicinalProduct>, List<MpdMedicinalProduct>> {
        val byId  = if (ids.isNotEmpty()) medicinalProductRepo.findAllByIdIn(ids) else emptyList()
        val byReg = if (regNumbers.isNotEmpty()) medicinalProductRepo.findAllByRegistrationNumberIn(regNumbers) else emptyList()
        val all   = (byId + byReg).distinctBy { it.id }

        return if (unitMode == MedicinalUnitMode.DAILY_DOSES) {
            all.partition { prod ->
                prod.dailyDosePackaging != null && prod.dailyDosePackaging > BigDecimal.ZERO
            }
        } else {
            all to emptyList()
        }
    }

    private fun List<MpdMedicinalProduct>.toDto() =
        map { MedicinalProductIdentificators(it.id!!, it.suklCode) }
}
